source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,315 @@
+#include "common/network/apple_dns_impl.h"
+
+#include <dns_sd.h>
+
+#include <chrono>
+#include <cstdint>
+#include <list>
+#include <memory>
+#include <string>
+
+#include "envoy/common/platform.h"
+
+#include "common/common/assert.h"
+#include "common/common/fmt.h"
+#include "common/network/address_impl.h"
+#include "common/network/utility.h"
+
+#include "absl/strings/str_join.h"
+
+namespace Envoy {
+namespace Network {
+
+AppleDnsResolverImpl::AppleDnsResolverImpl(Event::Dispatcher& dispatcher)
+    : dispatcher_(dispatcher) {
+  ENVOY_LOG(warn, "Constructing DNS resolver");
+  initializeMainSdRef();
+}
+
+AppleDnsResolverImpl::~AppleDnsResolverImpl() {
+  ENVOY_LOG(warn, "Destructing DNS resolver");
+  deallocateMainSdRef();
+}
+
+void AppleDnsResolverImpl::deallocateMainSdRef() {
+  ENVOY_LOG(warn, "DNSServiceRefDeallocate main sd ref");
+  // dns_sd.h says:
+  //   If the reference's underlying socket is used in a run loop or select() call, it should
+  //   be removed BEFORE DNSServiceRefDeallocate() is called, as this function closes the
+  //   reference's socket.
+  sd_ref_event_.reset();
+  DNSServiceRefDeallocate(main_sd_ref_);
+}
+
+void AppleDnsResolverImpl::initializeMainSdRef() {
+  // This implementation uses a shared connection for two main reasons:
+  //    1. Efficiency of concurrent resolutions by sharing the same underlying UDS to the DNS
+  //       daemon.
+  //    2. An error on a connection to the daemon is good indication that other connection,
+  //       even if not shared, would not succeed. So it is better to share one connection and
+  //       promptly cancel all outstanding queries, rather than individually wait for all
+  //       connections to error out.
+  // However, using a shared connection brings some complexities detailed in the inline comments
+  // for kDNSServiceFlagsShareConnection in dns_sd.h, and copied (and edited) in this implementation
+  // where relevant.
```

Comment:
I don't feel strongly about it, but is this a premature optimization? Should we just do a connection per resolution for now and keep it simple?

Reply:
I ultimately decided to leave it. I think the reasons in the inline comment are strong enough to leave it. Mostly around "failing fast" in mobile environments were instead of potentially waiting for independent timeouts that may stagger, all outstanding queries will fail.